ch_util/cal_utils.py

```diff
@@ -2161,3 +2166,315 @@ def _dec_to_el(dec):
+        "vira": ephemeris.VirA,
+    }
+
+    # Get the set of gain IDs for each time stamp
+    gain_ids = state_id_of_type(dataset_ids, "gains")
+    collapsed_ids = unique_unmasked_entry(gain_ids, axis=0)
+    unique_gains_ids = np.unique(collapsed_ids.compressed())
+
+    gain_info_dict = {}
+
+    # For each gain update extract all the relevant information
+    for state_id in unique_gains_ids:
+
+        d = {}
+        gain_info_dict[state_id] = d
+
+        # Extract the update ID
+        update_id = ds.DatasetState.from_id(state_id).data["data"]["update_id"]
+
+        # Parse the ID for the required information
+        split_id = update_id.split("_")
+        # After restart we sometimes have only a timing update without a source
+        # reference. These aren't valid for our purposes here, and can be distinguished
+        # at the update_id doesn't contain source information, and is thus shorter
+        d["valid"] = len(split_id) > 3
+
+        # If it's not a valid update we shouldn't try to extract everything else
+        if not d["valid"]:
+            continue
+
+        d["gen_time"] = ctime.datetime_to_unix(ctime.timestr_to_datetime(split_id[1]))
+        d["interpolated"] = "transition" in split_id
+        d["source_name"] = split_id[2].lower()
+
+        # Calculate the source transit time, and sanity check it
+        source = _source_dict[d["source_name"]]
+        d["source_transit"] = ephemeris.transit_times(
+            source, d["gen_time"] - 24 * 3600.0
+        )
+        cal_diff_hours = (d["gen_time"] - d["source_transit"]) / 3600
+        if cal_diff_hours > 3:
+            logger.warn(
+                f"Transit time ({ctime.unix_to_datetime(d['source_transit'])}) "
+                f"for source {d['source_name']} was a surprisingly long time "
+                f"before the gain update time ({cal_diff_hours} hours)."
+            )
+
+    # Array to store the extracted times in
+    reftime = np.zeros(len(collapsed_ids), dtype=np.float64)
+    reftime_prev = np.zeros(len(collapsed_ids), dtype=np.float64)
+    interp_start = np.zeros(len(collapsed_ids), dtype=np.float64)
+    interp_stop = np.zeros(len(collapsed_ids), dtype=np.float64)
+
+    # Iterate forward through the updates, setting transit times, and keeping track of
+    # the last valid update. This is used to set the previous source transit and the
+    # interpolation start time for all blended updates
+    last_valid_non_interpolated = None
+    for ii, state_id in enumerate(collapsed_ids):
+
+        valid_id = not np.ma.is_masked(state_id)
+        update = gain_info_dict[state_id] if valid_id else {}
+        valid = valid_id and update["valid"]
+
+        reftime[ii] = update["source_transit"] if valid else np.nan
+
+        if valid and update["interpolated"] and last_valid_non_interpolated is not None:
+            reftime_prev[ii] = reftime[last_valid_non_interpolated]
+            interp_start[ii] = times[last_valid_non_interpolated]
```

Comment:
```suggestion
            interp_start[ii] = times[last_non_interpolated]
```
As discussed in my summary, this would start interpolating from the last non-interpolated update, whether or not it is valid.